proto/background.go

```diff
@@ -46,19 +45,25 @@ func (a *Agent) startHeartbeat(ctx context.Context, tc *taskContext, heartbeat c
 	}
 }
 
-func (a *Agent) startIdleTimeoutWatch(ctx context.Context, tc *taskContext, idleTimeout chan<- struct{}, resetIdleTimeout <-chan time.Duration) {
+func (a *Agent) startIdleTimeoutWatch(ctx context.Context, tc *taskContext, idleTimeout chan<- struct{}, resetIdleTimeout chan time.Duration) {
 	idleTimeoutInterval := defaultIdleTimeout
 	if a.opts.IdleTimeoutInterval != 0 {
 		idleTimeoutInterval = a.opts.IdleTimeoutInterval
 	}
 	timer := time.NewTimer(idleTimeoutInterval)
+
 	defer timer.Stop()
 	for {
 		select {
 		case <-timer.C:
-			tc.logger.Execution().Error("Hit idle timeout")
-			close(idleTimeout)
-			return
+			// check the time of the last successful
+			// request against the API. if i'
+			if time.Now().Sub(a.comm.LastMessageAt()) > idleTimeoutInterval {
+				tc.logger.Execution().Error("Hit idle timeout")
+				close(idleTimeout)
+				return
+			}
+			a.checkIn(ctx, tc, tc.currentCommand, idleTimeoutInterval, resetIdleTimeout)
```

Comment:
This should reset the timeout to `idleTimeoutInterval.Sub(time.Now().Sub(a.comm.LastMessageAt()))` (with error checking that it's not negative, or save time.Now() on 61).

Reply:
updated